lib/graphql/query_resolver.rb

```diff
@@ -3,75 +3,92 @@
 
 module GraphQL
   module QueryResolver
+    def self.run(model_class, context, _return_type, &block)
+      Resolver.new(model_class, context).call(&block)
+    end
 
-    def self.run(model_class, context, return_type)
-      to_load = yield
-      dependencies = {}
-
-      reflection_dependencies = map_dependencies(model_class, context.ast_node)
-      dependencies = reflection_dependencies.merge(dependencies)
+    class Resolver
+      attr_reader :context, :fragment_definitions
 
-      if dependencies.any? && to_load.present?
-        if ActiveRecord::VERSION::MAJOR < 4
-          ActiveRecord::Associations::Preloader.new(to_load, dependencies).run
-        else
-          ActiveRecord::Associations::Preloader.new.preload(to_load, dependencies)
-        end
+      def initialize(model_class, context)
+        @model_class = model_class
+        @context = context
+        @fragment_definitions = context.query.fragments
       end
 
-      to_load
-    end
+      def call
+        to_load = yield
+        dependencies = map_dependencies(@model_class, context.ast_node)
 
-    def self.using_relay_pagination?(selection)
-      selection.name == 'edges'
-    end
+        if dependencies.any? && to_load.present?
+          preload_dependencies(to_load, dependencies)
+        end
 
-    def self.using_nodes_pagination?(selection)
-      selection.name == 'nodes'
-    end
+        to_load
+      end
 
-    def self.map_relay_pagination_depencies(class_name, selection, dependencies)
-      node_selection = selection.selections.find { |sel| sel.name == 'node' }
+      private
 
-      if node_selection.present?
-        map_dependencies(class_name, node_selection, dependencies)
-      else
-        dependencies
-      end
-    end
+      def preload_dependencies(to_load, dependencies)
+        if ActiveRecord::VERSION::MAJOR < 4
+          return ActiveRecord::Associations::Preloader.new(
+            to_load, dependencies
+          ).run
+        end
 
-    def self.has_reflection_with_name?(class_name, selection_name)
-      class_name.reflections.with_indifferent_access[selection_name].present?
-    end
+        ActiveRecord::Associations::Preloader.new.preload(to_load, dependencies)
+      end
 
-    def self.map_dependencies(class_name, ast_node, dependencies={})
-      ast_node.selections.each do |selection|
-        name = selection.name
+      def map_dependencies(class_name, ast_node, dependencies = {})
+        ast_node.selections.each do |selection|
+          if inline_fragment?(selection) || relay_connection_using_nodes?(selection)
+            dig_deeper = selection
```

Comment:
`dig_deeper` sounds a little weird. But I can't quite think of an alternative name here...

Reply:
it is almost like: "we want to replace the current selection with this nested option".